dkg-gadget/src/keygen_manager/mod.rs

```diff
@@ -0,0 +1,385 @@
+use crate::{
+	async_protocols::{remote::AsyncProtocolRemote, KeygenPartyId, KeygenRound},
+	gossip_engine::GossipEngineIface,
+	signing_manager::work_manager::{JobMetadata, PollMethod, WorkManager},
+	utils::SendFuture,
+	worker::{DKGWorker, HasLatestHeader, KeystoreExt, ProtoStageType},
+	Client,
+};
+use atomic::Atomic;
+use dkg_logging::debug_logger::DebugLogger;
+use dkg_primitives::types::{DKGError, SignedDKGMessage};
+use dkg_runtime_primitives::{
+	crypto::{AuthorityId, Public},
+	keccak_256, DKGApi, MaxAuthorities, MaxProposalLength, SessionId, GENESIS_AUTHORITY_SET_ID,
+};
+use sc_client_api::Backend;
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::traits::{Block, Header, NumberFor};
+use std::{
+	marker::PhantomData,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicUsize, Ordering},
+		Arc,
+	},
+};
+
+pub struct KeygenManager<B: Block, BE, C, GE> {
+	// governs the workload for each node
+	work_manager: WorkManager<B>,
+	active_keygen_retry_id: Arc<AtomicUsize>,
+	keygen_state: Arc<Atomic<KeygenState>>,
+	latest_executed_session_id: Arc<Atomic<Option<SessionId>>>,
+	pub finished_count: Arc<AtomicUsize>,
+	_pd: PhantomData<(B, BE, C, GE)>,
+}
+
+impl<B: Block, BE, C, GE> Clone for KeygenManager<B, BE, C, GE> {
+	fn clone(&self) -> Self {
+		Self {
+			work_manager: self.work_manager.clone(),
+			_pd: self._pd,
+			active_keygen_retry_id: self.active_keygen_retry_id.clone(),
+			keygen_state: self.keygen_state.clone(),
+			latest_executed_session_id: self.latest_executed_session_id.clone(),
+			finished_count: self.finished_count.clone(),
+		}
+	}
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum KeygenState {
+	Uninitialized,
+	RunningKeygen,
+	RunningGenesisKeygen,
+	// session_completed denotes the session that executed the keygen, NOT
+	// the generated DKG public key for the next session
+	KeygenCompleted { session_completed: u64 },
+	Failed { session_id: u64 },
+}
+
+// only 1 task at a time may run for keygen
+const MAX_RUNNING_TASKS: usize = 1;
```

Comment:
For these parameters is there any way we can push it out to the CLI/WorkerParams?